### DISCUSSION TITLE
What is the recommended Jotai store layout?

### DISCUSSION THREAD

**Discussion by AjaxSolutions** [pmndrs/jotai#1884] [Q&A]
I know Jotai is very unopinionated, but what is the recommended Jotai store layout, folder structure, etc.

I'm curious what you think about the advantages and disadvantages of having one atom per file?

How do you organize atoms, hooks and actions in a Jotai store?

**Comment by devnomic** [comment]
This is my opinion after building with jotai:

Every components should have their atom files. But you may have global store for global atoms.

The structure

````
src/components/SignInModal/index.tsx <- UI component
src/components/SignInModal/store.ts <- contains atoms for sign in modal (visibleAtom, showAtom, hideAtom, titleAtom, etc)
````
Or you may want to put both atoms and it into one component.
The advantage is huge, you don't need to search which atom handle which component, and less asking where is the logic for this component.

For global atoms
````
src/stores/account.ts

````

Also from my experience, most likely you will write less hooks (less code = less bugs) .

For example, lets say you have a sign in modal,
You may tempted to write hook for it such as
````
const { show } = useSignInModal(); 
````
and inside the hook probably you are going to update jotai store from there
````
export function useSignInModal() {
   const [visible, setVisible] = useAtom(visibleAtom);

   function show() {
     setVisible(true);
   }

  return {
    visible,
    show,
  }
}
````
While the code looks good, in jotai this will cause unnecessary re-render if you only need to call `show`.

That's why it's better create the atom and access it directly (also less code).
````
export visibleAtom = atom(false);
export showAtom = atom(null, (_, set) => set(visibleAtom, true));

function Component() {
   const show = useSetAtom(showAtom);
}
````
